### PR TITLE
add adjacency operator.

### PR DESCRIPTION
**Pull request by BullDemonKing**
## Description ##
The adjacency operator is implemented for DGL that converts a CSR matrix with edge Id to an adjacency matrix for SpMV.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by stu1130** [comment]
@mxnet-label-bot add [pr-awaiting-review]
Thanks! @BullDemonKing

File: src/operator/contrib/dgl_graph.cc [hunk 1]
```diff
@@ -459,6 +456,69 @@ The storage type of ``edge_id`` output depends on storage types of inputs
...
+  if (!dispatched) {
+    LOG(ERROR) << "Cannot dispatch edge_id storage type, only works for csr matrices";
+  }
+  return dispatched;
+}
+
+NNVM_REGISTER_OP(_contrib_dgl_adjacency)
+.describe(R"code(This operator converts a CSR matrix whose values are edge Ids
+to an adjacency matrix whose values are ones.
+Example::
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: Add an empty newline after Example::

File: src/operator/contrib/dgl_graph.cc [hunk 2]
```diff
@@ -459,6 +456,69 @@ The storage type of ``edge_id`` output depends on storage types of inputs
...
+                                    DispatchMode* dispatch_mode,
+                                    std::vector<int>* in_attrs,
+                                    std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1U) << "Only works for 2d arrays";
+  CHECK_EQ(out_attrs->size(), 1U);
+  int& out_stype = out_attrs->at(0);
+  bool dispatched = storage_type_assign(&out_stype, kCSRStorage,
+                                        dispatch_mode, DispatchMode::kFComputeEx);
+  if (!dispatched) {
+    LOG(ERROR) << "Cannot dispatch edge_id storage type, only works for csr matrices";
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: edge_id  -> dgl_adjacency

Comment (eric-haibin-lin) [reply src/operator/contrib/dgl_graph.cc]: Also log the value for out_stype using `common::stype_string`

File: src/operator/contrib/dgl_graph.cc [hunk 3]
```diff
@@ -459,6 +456,69 @@ The storage type of ``edge_id`` output depends on storage types of inputs
...
+to an adjacency matrix whose values are ones.
+Example::
+  x = [[ 1, 0, 0 ],
+       [ 0, 2, 0 ],
+       [ 0, 0, 3 ]]
+  dgl_adjacency(x) =
+      [[ 1, 0, 0 ],
+       [ 0, 1, 0 ],
+       [ 0, 0, 1 ]]
+
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph.cc]: Could you also document that the output dtype will always be float32?

File: tests/python/unittest/test_dgl_graph.py [hunk 4]
```diff
@@ -75,6 +75,16 @@ def test_subgraph():
             v2 = vertices[subv2]
             assert sp_g[v1, v2] == sp_subg[subv1, subv2]
 
+def test_adjacency():
+    sp_g, g = generate_graph(100)
+    start = time.time()
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_dgl_graph.py]: remove `time.time()`?

File: src/operator/contrib/dgl_graph-inl.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+void DGLAdjacencyForwardEx(const nnvm::NodeAttrs& attrs,
+                           const OpContext& ctx,
+                           const std::vector<NDArray>& inputs,
+                           const std::vector<OpReqType>& req,
+                           const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  CHECK_EQ(inputs[0].storage_type(), kCSRStorage);
+  CHECK_EQ(outputs[0].storage_type(), kCSRStorage);
```
Comment (eric-haibin-lin) [inline src/operator/contrib/dgl_graph-inl.h]: Check req == kWriteTo.

**Pull request merged**